app/src/main/java/com/weatherxm/usecases/ForecastUseCaseImpl.kt

```diff
@@ -0,0 +1,63 @@
+package com.weatherxm.usecases
+
+import arrow.core.Either
+import com.weatherxm.data.ApiError
+import com.weatherxm.data.Failure
+import com.weatherxm.data.HourlyWeather
+import com.weatherxm.data.network.ErrorResponse.Companion.INVALID_TIMEZONE
+import com.weatherxm.data.repository.WeatherForecastRepository
+import com.weatherxm.ui.common.UIDevice
+import com.weatherxm.ui.common.UIForecast
+import com.weatherxm.ui.common.UIForecastDay
+import java.time.ZoneId
+import java.time.ZonedDateTime
+
+@Suppress("LongParameterList")
+class ForecastUseCaseImpl(
+    private val weatherForecastRepository: WeatherForecastRepository
+) : ForecastUseCase {
+
+    @Suppress("MagicNumber")
+    override suspend fun getForecast(
+        device: UIDevice,
+        forceRefresh: Boolean
+    ): Either<Failure, UIForecast> {
+        if (device.timezone.isNullOrEmpty()) {
+            return Either.Left(ApiError.UserError.InvalidTimezone(INVALID_TIMEZONE))
+        }
+        val nowDeviceTz = ZonedDateTime.now(ZoneId.of(device.timezone))
+        val dateEndInDeviceTz = nowDeviceTz.plusDays(7)
+        return weatherForecastRepository.getDeviceForecast(
+            device.id,
+            nowDeviceTz,
+            dateEndInDeviceTz,
+            forceRefresh
+        ).map { result ->
+            val nextHourlyWeatherForecast = mutableListOf<HourlyWeather>()
+            val forecastDays = result.map { weatherData ->
+                weatherData.hourly?.filter {
+                    val isCurrentHour = it.timestamp.dayOfYear == nowDeviceTz.dayOfYear
+                        && it.timestamp.hour == nowDeviceTz.hour
+                    isCurrentHour || it.timestamp.isAfter(nowDeviceTz)
```

Comment:
Can we do something like the following?
```
it.timestamp >= nowDeviceTz && it.timestamp < nowDeviceTz.plusHours(24)
```
And omit picking the first 24 items. With this, we handle the case of gaps (if there are for any reason) and the case of non-hourly samples

Reply:
Valid point, and it's much cleaner now. Thanks!